File: c#/ElegantCode/ElegantCode/ElegantCode.Fundamental/Core/DriverAdapter/DriverAdapter.cs
using ElegantCode.Fundamental.Core.Entities;
using ElegantCode.Fundamental.Core.Errors;
using ElegantCode.Fundamental.Core.Presenter;
using ElegantCode.Fundamental.Core.UsesCases;

namespace ElegantCode.Fundamental.Core.DriverAdapter;

public static class DriverAdapter
{
    public static async Task<(Tout Entity, Error Error)> CreateUseCaseWorkflow<Tout, TUseCaseQuery, TUseCaseResult>(
        IValidateRequest<TUseCaseQuery> aRequestForDriverAdapter,
        IUseCaseAsync<TUseCaseQuery, TUseCaseResult> myUseCase,
        IPresenter<TUseCaseResult, Tout> presenter,
        CancellationToken cancellation = default)
        where TUseCaseQuery : IGotCorrelationToken
    {
        var validationResult = aRequestForDriverAdapter.ValidateRequest();
        presenter.PresentError(validationResult.Error);

        if (validationResult.IsNotOnError())
            await ExecuteUseCase(myUseCase, presenter, validationResult, cancellation);

        return await presenter.View();
    }

    public static async Task ExecuteUseCase<Tout, TUseCaseQuery, TUseCaseResult>(
        IUseCaseAsync<TUseCaseQuery, TUseCaseResult> myUseCAse,
        IPresenter<TUseCaseResult, Tout> doExemplePresenter,
        (TUseCaseQuery UseCaseQuery, Error Error) validationResult,
        CancellationToken cancellation = default)
        where TUseCaseQuery : IGotCorrelationToken
    {
        var useCaseQuery = validationResult.UseCaseQuery;
        try
        {
            doExemplePresenter.PresentData(await myUseCAse.Execute(useCaseQuery, cancellation));
        }
        catch (UseCaseException ex)
        {
            doExemplePresenter.PresentError(new(useCaseQuery.CorrelationToken, ex.Message));
        }
    }
}

File: TypeScript/Rpg/src/clean_archi/adapters/DriverAdapter.ts
import { CorrelationError, isOnError } from "../errors/CorrelationError";
import { IMUseCase } from "../IMUseCase";
import { IMPresenter } from "../IMPresenter";
import { PresentData } from "../PresentData";
import { IValidateRequest } from "../IValidateRequest";
import { IGotCorrelationToken } from "../IGotCorrelationToken";

export class DriverAdapter {
  public static async CreateUseCaseWorkflow<
    TOut,
    TUseCaseQuery extends IGotCorrelationToken,
    TUseCaseResult
  >(
    aRequestForDriverAdapter: IValidateRequest<TUseCaseQuery>,
    useCase: IMUseCase<TUseCaseQuery, TUseCaseResult>,
    presenter: IMPresenter<TUseCaseResult, TOut>
  ): Promise<PresentData<TOut>> {
    const query = aRequestForDriverAdapter.ValidateRequest();
    presenter.presentError(query[1]);

    if (isOnError(query[1]) === false) {
      await DriverAdapter.executeUseCase(
        useCase,
        query,
        presenter,
        aRequestForDriverAdapter
      );
    }

    return await presenter.view();
  }

  private static async executeUseCase<
    TOut,
    TUseCaseQuery extends IGotCorrelationToken,
    TUseCaseResult
  >(
    useCase: IMUseCase<TUseCaseQuery, TUseCaseResult>,
    query: [useCaseQuery: TUseCaseQuery, error: CorrelationError],
    presenter: IMPresenter<TUseCaseResult, TOut>,
    aRequestForDriverAdapter: IValidateRequest<TUseCaseQuery>
  ) {
    const useCaseQuery = query[0];
    try {
      await useCase.Execute(useCaseQuery, presenter);
    } catch (error) {
      presenter.presentError(
        new CorrelationError(aRequestForDriverAdapter.CorrelationToken, [
          error.message
        ])
      );
    }
  }
}
